load_test/lib/load_test/user/sse.ex

```diff
@@ -73,71 +91,106 @@ defmodule SseUser do
       sse_timeout: context.sse_timeout,
       start_publisher_callback: fn ->
         LoadTest.Main.start_publisher(context, user_name, topic, expected_messages)
-      end
+      end,
+      connect_callback: connect_callback,
+      last_event_id: nil,
+      reconnect: reconnect
     }
 
-    wait_for_messages(state, conn_pid, stream_ref, expected_messages)
+    state = connect_callback.(state)
+
+    wait_for_messages(state, expected_messages)
   end
 
-  defp wait_for_messages(state, conn_pid, stream_ref, [first_message | remaining_messages]) do
+  defp reconnect(state, messages, reason) do
+    :ok = :gun.close(state.conn_pid)
+
+    if state.reconnect == -1 do
+      Logger.error(fn -> "#{header(state)} Connection closed: #{reason}" end)
+      Stats.inc_msg_received_http_error()
+      raise("#{header(state)} Connection closed")
+    end
+
+    Logger.error("#{header(state)} Connection closed, reconnecting: #{reason}")
+    Stats.inc_reconnect()
+    state = state.connect_callback.(state)
+    state = Map.put(state, :reconnect, state.reconnect + 1)
+    wait_for_messages(state, messages)
+  end
+
+  defp wait_for_messages(state, [first_message | remaining_messages]) do
     Logger.debug(fn -> "#{header(state)} Waiting for message: #{first_message}" end)
 
-    result = :gun.await(conn_pid, stream_ref, state.sse_timeout)
+    result = :gun.await(state.conn_pid, state.stream_ref, state.sse_timeout)
 
     case result do
       {:response, _, code, _} when code == 200 ->
         Logger.debug(
           "#{header(state)} Connected, waiting: #{length(remaining_messages) + 1} messages, url #{state.url}"
         )
 
-        state.start_publisher_callback.()
-        wait_for_messages(state, conn_pid, stream_ref, [first_message | remaining_messages])
+        if state.start_publisher_callback != nil do
+          state.start_publisher_callback.()
+          state = Map.put(state, :start_publisher_callback, nil)
+          wait_for_messages(state, [first_message | remaining_messages])
+        else
+          wait_for_messages(state, [first_message | remaining_messages])
+        end
```

Comment:
```suggestion
        state =
          update_in(state.start_publisher_callback, fn callback ->
            if callback, do: (callback.(); nil), else: nil
          end)
        
        wait_for_messages(state, [first_message | remaining_messages])
```

Reply:
Does not work with struct. Improved the code anyway